apps/issuance/src/issuance.service.ts

```diff
@@ -425,259 +426,305 @@ export class IssuanceService {
-            };
-          }
-          
+    const issuanceMethodLabel = 'create-offer-oob';
+    const url = await this.getAgentUrl(
+      issuanceMethodLabel,
+      orgAgentType,
+      agentDetails.agentEndPoint,
+      agentDetails.tenantId
+    );
+    const organizationDetails = await this.issuanceRepository.getOrganization(orgId);
 
-          this.logger.log(`outOfBandIssuancePayload ::: ${JSON.stringify(outOfBandIssuancePayload)}`);
+    if (!organizationDetails) {
+      throw new NotFoundException(ResponseMessages.issuance.error.organizationNotFound);
+    }
 
-          const credentialCreateOfferDetails = await this._outOfBandCredentialOffer(outOfBandIssuancePayload, url, apiKey);
+    let apiKey: string = await this.cacheService.get(CommonConstants.CACHE_APIKEY_KEY);
+    if (!apiKey || null === apiKey || undefined === apiKey) {
+      apiKey = await this._getOrgAgentApiKey(orgId);
+    }
 
-          if (!credentialCreateOfferDetails) {
-            errors.push(new NotFoundException(ResponseMessages.issuance.error.credentialOfferNotFound));
-            return false;
-          }
+    const errors = [];
+    const emailPromises = [];
+    const sendEmailCredentialOffer: {
+      iterator: CredentialOffer;
+      emailId: string;
+      index: number;
+      credentialType: IssueCredentialType;
+      protocolVersion: string;
+      attributes: IAttributes[];
+      credentialDefinitionId: string;
+      outOfBandCredential: OutOfBandCredentialOfferPayload;
+      comment: string;
+      organisation: organisation;
+      errors: string[];
+      url: string;
+      apiKey: string;
+      organizationDetails: organisation;
+    } = {
+      credentialType,
+      protocolVersion,
+      attributes,
+      credentialDefinitionId,
+      outOfBandCredential,
+      comment,
+      organisation,
+      errors,
+      url,
+      apiKey,
+      organizationDetails,
+      iterator: undefined,
+      emailId: emailId || '',
+      index: 0
+    };
 
-          const invitationId = credentialCreateOfferDetails.response.invitation['@id'];
+    if (credentialOffer) {
+      for (let i = 0; i < credentialOffer.length; i += Number(process.env.OOB_BATCH_SIZE)) {
+        const batch = credentialOffer.slice(i, i + Number(process.env.OOB_BATCH_SIZE));
+        // Process each batch in parallel
+        const batchPromises = batch.map(async (iterator, index) => {
+          
+          sendEmailCredentialOffer['iterator'] = iterator;
+          sendEmailCredentialOffer['emailId'] = iterator.emailId;
+          sendEmailCredentialOffer['index'] = index;
 
-          if (!invitationId) {
-            errors.push(new NotFoundException(ResponseMessages.issuance.error.invitationNotFound));
-            return false;
-          }
+          return this.sendEmailForCredentialOffer(sendEmailCredentialOffer);
+        });
+        emailPromises.push(Promise.all(batchPromises));
```

Comment:
I am not sure how it will handle the remaining batch if the 1st batch will fail.

Reply:
We will check the possibilities for validation and we will create separate PR for it.